src/libraries/Common/src/System/Net/NetworkInformation/InterfaceInfoPal.Unix.cs

```diff
@@ -1,13 +1,76 @@
+
 namespace System.Net.NetworkInformation
 {
     internal static class InterfaceInfoPal
     {
-        public static uint InterfaceNameToIndex(string interfaceName)
+        // Measured in bytes (the native data type for this platform's API.)
+        // Selected to occupy as much stack space as Windows' threshold (which is 256 two-byte characters.)
+        private const int StackAllocationThreshold = 512;
+
+        public static unsafe uint InterfaceNameToIndex<TChar>(ReadOnlySpan<TChar> interfaceName)
+            where TChar : unmanaged, IBinaryNumber<TChar>
         {
-            return Interop.Sys.InterfaceNameToIndex(interfaceName);
+            Debug.Assert(typeof(TChar) == typeof(byte) || typeof(TChar) == typeof(char));
+
+            // Measured in bytes, including null terminator.
+            int bufferSize = 0;
+            byte* nativeMemory = null;
+
+            // The underlying API for this method accepts a null-terminated UTF8 string containing the interface name.
+            // If TChar is byte, the only work required is a byte copy. If TChar is char, there's a transcoding step from Unicode
+            // to UTF8.
+            // Notably: the encoding passed to the underlying API is different between Linux and Windows.
+            if (typeof(TChar) == typeof(byte))
+            {
+                bufferSize = interfaceName.Length + 1;
+            }
+            else if (typeof(TChar) == typeof(char))
+            {
+                ReadOnlySpan<char> castInterfaceName = MemoryMarshal.Cast<TChar, char>(interfaceName);
+
+                bufferSize = Encoding.UTF8.GetByteCount(castInterfaceName) + 1;
+            }
+            Debug.Assert(bufferSize <= int.MaxValue - 1);
+
+            try
+            {
+                nativeMemory = bufferSize <= StackAllocationThreshold ? null : (byte*)NativeMemory.Alloc((nuint)bufferSize);
+
+                Span<byte> buffer = nativeMemory == null ? stackalloc byte[StackAllocationThreshold].Slice(0, bufferSize) : new Span<byte>(nativeMemory, bufferSize);
+
+                if (typeof(TChar) == typeof(byte))
+                {
+                    ReadOnlySpan<byte> castInterfaceName = MemoryMarshal.Cast<TChar, byte>(interfaceName);
+
+                    castInterfaceName.CopyTo(buffer);
+                }
+                else if (typeof(TChar) == typeof(char))
+                {
+                    ReadOnlySpan<char> castInterfaceName = MemoryMarshal.Cast<TChar, char>(interfaceName);
+
+                    Encoding.UTF8.GetBytes(castInterfaceName, buffer);
+                }
+                buffer[buffer.Length - 1] = 0;
+
+                return Interop.Sys.InterfaceNameToIndex(buffer);
+            }
+            finally
+            {
+                if (nativeMemory != null)
+                {
+                    int errNo = Marshal.GetLastPInvokeError();
+
+                    NativeMemory.Free(nativeMemory);
+                    Marshal.SetLastPInvokeError(errNo);
```

Comment:
You're resetting the pinvoke error here, but not in the Windows PAL. Was that intentional?

Reply:
It's intentional - the Unix PAL method is [`SystemNative_InterfaceNameToIndex`](https://github.com/dotnet/runtime/blob/main/src/native/libs/System.Native/pal_networking.c#L3678), and this calls `if_nametoindex`, which sets errno. We preserve this to protect it from the impact of `NativeMemory.Free`.

The Windows PAL calls [`ConvertInterfaceNameToLuidW`](https://learn.microsoft.com/en-us/windows/win32/api/netioapi/nf-netioapi-convertinterfacenametoluidw), and this doesn't change the return value of GetLastError.